### PR TITLE
handle valid moves and attacking peices 

### PR DESCRIPTION
**Pull request by PeteMango**
- refactor piece attacking to be handled by piece class in a non virtual function
- added move validation for all the pieces
- tested with piece setup and printing

### CODE REVIEW

**Comment by PeteMango** [comment]
sample output: 

```
root@a6efe98e84e8:/usr/src/app# ./out/chess 
setup
+ K c6
can attack: 
5b
5c
5d
6b
6d
7b
7c
7d
8  _ _ _ _
7 _ _ _ _ 
6  _K_ _ _
5 _ _ _ _ 
4  _ _ _ _
3 _ _ _ _ 
2  _ _ _ _
1 _ _ _ _ 

  abcdefgh
+ Q b1
can attack: 
1a
1c
1d
1e
1f
1g
1h
2a
2b
2c
3b
3d
4b
4e
5b
5f
6b
6g
7b
7h
8b
8  _ _ _ _
7 _ _ _ _ 
6  _K_ _ _
5 _ _ _ _ 
4  _ _ _ _
3 _ _ _ _ 
2  _ _ _ _
1 _Q_ _ _ 

  abcdefgh
```

**Comment by PeteMango** [comment]
shows how 3c and 3e were moves that we can attack but not moves we can move to
```
setup
= black
+ P d3
can attack: 
2c
2e
can move: 
2d
8  _ _ _ _
7 _ _ _ _ 
6  _ _ _ _
5 _ _ _ _ 
4  _ _ _ _
3 _ _p_ _ 
2  _ _ _ _
1 _ _ _ _ 

  abcdefgh
= white
+ P d2
can attack: 
3c
3e
can move: 
8  _ _ _ _
7 _ _ _ _ 
6  _ _ _ _
5 _ _ _ _ 
4  _ _ _ _
3 _ _p_ _ 
2  _ P _ _
1 _ _ _ _ 

  abcdefgh
  ```

File: include/piece/king.h [hunk 1]
```diff
@@ -5,17 +5,13 @@
...
-    std::vector<std::pair<int, int>> directions
-        = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 }, { 1, 1 },
-              { 1, -1 }, { -1, 1 }, { -1, -1 } };
 
 public:
     King(Color color, Coordinate location, PieceType type,
         std::shared_ptr<Board> board);
-    bool is_valid_move(Coordinate square) override;
-    bool is_check(std::set<Coordinate>& attacked_squares);
+    void get_threatened_squares(std::set<Coordinate>& s) override;
```
Comment (normanchenn) [inline include/piece/king.h:12]: do you need to do override get valid moves

File: include/piece/rook.h [hunk 2]
```diff
@@ -4,14 +4,12 @@
 #include "piece.h"
 
 class Rook : public Piece {
-    std::vector<std::pair<int, int>> directions
-        = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };
 
 public:
+    bool castleable;
```
Comment (normanchenn) [inline include/piece/rook.h:9]: we maintain this in castle rights

File: src/piece/pawn.cpp [hunk 3]
```diff
@@ -1,26 +1,54 @@
...
-void Pawn::get_attacking_squares(std::set<Coordinate>& s)
+void Pawn::get_valid_moves(std::set<Coordinate>& s)
 {
+    /* handle captures */
+    for (auto i : this->captures) {
+        std::pair<int, int> start = get_grid_indexes(this->location);
+        start = add_pairs(start, i);
+
+        if (coordinate_in_bounds(start)
+            and this->board->get_grid()[start.first][start.second]
```
Comment (normanchenn) [inline src/piece/pawn.cpp:25]: en passant?

File: src/piece/pawn.cpp [hunk 4]
```diff
@@ -1,26 +1,54 @@
...
+        }
+    }
+    /* handles moving forward */
     for (auto i : this->directions) {
-        std::pair<int, int> start
-            = get_grid_indexes(this->location);
+        std::pair<int, int> start = get_grid_indexes(this->location);
         start = add_pairs(start, i);
+
+        if (coordinate_in_bounds(start)
```
Comment (normanchenn) [inline src/piece/pawn.cpp:36]: double moves for first move?

File: src/piece/piece.cpp [hunk 5]
```diff
@@ -1,15 +1,102 @@
...
+void Piece::filter_moves(std::set<Coordinate>& s)
+{
+    std::set<Coordinate> good;
+    for (const Coordinate& c : s) {
+        std::pair<int, int> index
+            = get_grid_indexes(c); /* filter out the piece that coordinates that
+                                      already has an allied piece */
+        if (!this->board->get_grid()[index.first][index.second]
+            or this->board->get_grid()[index.first][index.second]
+                and this->board->get_grid()[index.first][index.second]->get_color()
```
Comment (normanchenn) [inline src/piece/piece.cpp:85]: add brackets?

**Review by normanchenn — CHANGES_REQUESTED** [review]

File: src/piece/piece.cpp (hunk 5 above)
Comment (PeteMango) [reply src/piece/piece.cpp:85]: this is the formatter's work

File: src/piece/pawn.cpp (hunk 4 above)
Comment (PeteMango) [reply src/piece/pawn.cpp:36]: resolved.

File: src/piece/pawn.cpp (hunk 3 above)
Comment (PeteMango) [reply src/piece/pawn.cpp:25]: will add this functionality later.

File: include/piece/king.h (hunk 1 above)
Comment (PeteMango) [reply include/piece/king.h:12]: resolved.

File: include/piece/rook.h (hunk 2 above)
Comment (PeteMango) [reply include/piece/rook.h:9]: resolved.

**Review by normanchenn — APPROVED** [review]

**Review by Chen-Prototypes — APPROVED** [review]

**Pull request merged**